### PR TITLE
Use JONI to replace java.regex for regex sql functions

### PR DESCRIPTION
**Pull request by haozhun**
java.regex

    (patternString)  (sourceLength)  Mode  Samples           Score    Score error  Units
              .*x.*            1024  avgt       10     3330533.930     287806.854  ns/op
              .*x.*           32768  avgt       10  3396127115.000  242129758.239  ns/op
          .*(x|y).*            1024  avgt       10     6847263.706     630131.295  ns/op
          .*(x|y).*           32768  avgt       10  7405041415.600  481333551.769  ns/op
        longdotstar            1024  avgt       10     4386599.874     269038.787  ns/op
        longdotstar           32768  avgt       10  5272199253.700  563624383.505  ns/op
              phone            1024  avgt       10       13611.496       1390.554  ns/op
              phone           32768  avgt       10       22267.639       3061.627  ns/op
            literal            1024  avgt       10        2225.823        285.027  ns/op
            literal           32768  avgt       10       65688.858       4443.133  ns/op

joni

    (patternString)  (sourceLength)  Mode  Samples       Score  Score error  Units
              .*x.*            1024  avgt       10    5929.790      103.427  ns/op
              .*x.*           32768  avgt       10  189781.419     4770.151  ns/op
          .*(x|y).*            1024  avgt       10    2764.053      143.002  ns/op
          .*(x|y).*           32768  avgt       10   94982.894     4222.652  ns/op
        longdotstar            1024  avgt       10     693.856       26.088  ns/op
        longdotstar           32768  avgt       10   20160.877      720.529  ns/op
              phone            1024  avgt       10    6841.006      274.788  ns/op
              phone           32768  avgt       10   12165.194      191.484  ns/op
            literal            1024  avgt       10    1183.754       70.801  ns/op
            literal           32768  avgt       10   36016.862      880.373  ns/op

### CODE REVIEW

**Comment by cberner** [comment]
It looks like there are some checkstyle failures

**Comment by haozhun** [comment]
Fixed checkstyle

**Comment by xerial** [comment]
@haozhun @cberner 
Could you evaluate this performance of this fix for multi-byte characters? We have encountered performance problem of joni when handling multi-byte (non-ascii) characters. Since joni doesn't accept invalid UTF-8 characters, presto-0.100's LikeFunction code introduces overhead for converting Slice -> Java String (UTF-8) -> getBytes(UTF-8).

In presto 0.100, LIKE '%(pattern)%' evaluation for non-ascii characters involves three memory copies:
```
   @ScalarFunction(value = "like", hidden = true)
    @SqlType(StandardTypes.BOOLEAN)
    public static boolean like(@SqlType(StandardTypes.VARCHAR) Slice value, @SqlType(LikePatternType.NAME) Regex pattern)
    {
        // Joni doesn't handle invalid UTF-8, so replace invalid characters
        byte[] bytes = value.getBytes();   <--  1st copy
        if (isAscii(bytes)) {
            return regexMatches(pattern, bytes);
        }
        // convert to a String and back to "fix" any broken UTF-8 sequences
        return regexMatches(pattern, value.toStringUtf8().getBytes(UTF_8)); <-- second, third copy
    }
```

For non-ascii characters, using `regexp_like(x, '(pattern)')` is 2x faster since it has only 1 memory copy for converting the string into UTF-8: 
```
    @Description("returns substrings matching a regular expression")
    @ScalarFunction
    @SqlType(StandardTypes.BOOLEAN)
    public static boolean regexpLike(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Pattern pattern)
    {
        return pattern.matcher(source.toString(UTF_8)).find();
    }
```

**Comment by haozhun** [comment]
@xerial 

The reason we needed to do that hack was because JONI does not handle error cases correctly when the underlying Character Encoding tells it that the input is invalid. That will result in infinite loop in JONI. Back when we implemented LIKE, we decided to make the memory copy hack you mentioned above. The performance penalty of the 3 memory copy applies to all characters, multi-byte or ascii.

This time, we figured several different ways to solve this issue. None of them involves massive memory copying. One of them is #2791 (WIP) which can do encoding sanity checking without copying. We are not using that here though because we figured that the above infinite loop issue can be mitigated by using a lenient variant of UTF8, which is simpler and more elegant.

Anyway, if you can provide me a multi-byte test case you have on hand that shows LIKE performs worse than regexp, I'd be glad to add it to my benchmark and see how this PR affects it.

**Comment by xerial** [comment]
Thanks for the clarification. Using a lenient mode for UTF8 looks promising.

Do you expect a benchmark like this commit? https://github.com/dain/presto/commit/087275a1fd489f7e79488fe0a1391f758ee56600

**Comment by haozhun** [comment]
@xerial  This PR already has a benchmark. But it may not be representative for the use cases you normally see. That's why I asked if you can provide a case.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/RegexpFunctionsBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+@BenchmarkMode(AverageTime)
+@Fork(1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+public class RegexpFunctionsBenchmark
+{
+    @Benchmark
+    public boolean benchmarkLike(DotStarAroundData data)
+    {
+        return RegexpFunctions.regexpLike(data.getSource(), data.getPattern());
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/RegexpFunctionsBenchmark.java]: static import

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 2]
```diff
@@ -78,17 +85,118 @@ public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source,
...
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacement.toStringUtf8());
+                }
+                nextByte = replacement.getByte(idx);
+                int backref;
+                if (nextByte == '{') { // case 1 in the above comment
+                    idx++;
+                    int startCursor = idx;
+                    while (idx < replacement.length()) {
+                        nextByte = replacement.getByte(idx);
+                        if (nextByte == '}') {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: If group name is unicode could this fail? Can the '}' byte be inside another codepoint?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 3]
```diff
@@ -78,17 +85,118 @@ public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source,
...
+                break;
+            }
+            if (matcher.getEnd() == matcher.getBegin()) {
+                nextStart = matcher.getEnd() + 1;
+            }
+            else {
+                nextStart = matcher.getEnd();
+            }
+            Slice sliceBetweenReplacements = source.slice(lastEnd, matcher.getBegin() - lastEnd);
+            lastEnd = matcher.getEnd();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: I found it confusing that we don't set `lastEnd = nextStart`, add some comments to this loop

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 4]
```diff
@@ -96,20 +204,37 @@ public static Slice regexpExtractAll(@SqlType(StandardTypes.VARCHAR) Slice sourc
...
-        Matcher matcher = pattern.matcher(source.toString(UTF_8));
-        validateGroup(group, matcher);
-        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), 1024);
-        while (matcher.find()) {
-            String string = matcher.group(Ints.checkedCast(group));
-            if (string == null) {
+        Matcher matcher = pattern.matcher(source.getBytes());
+        validateGroup(group, matcher.getEagerRegion());
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 32);
+        int iGroup = Ints.checkedCast(group);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: This variable name is weird. I'd call it `group` and then rename `group` to `inputGroup`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 5]
```diff
@@ -126,36 +251,65 @@ public static Slice regexpExtract(@SqlType(StandardTypes.VARCHAR) Slice source,
...
     @SqlType(StandardTypes.VARCHAR)
-    public static Slice regexpExtract(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Pattern pattern, @SqlType(StandardTypes.BIGINT) long group)
+    public static Slice regexpExtract(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Regex pattern, @SqlType(StandardTypes.BIGINT) long group)
     {
-        Matcher matcher = pattern.matcher(source.toString(UTF_8));
-        validateGroup(group, matcher);
-        if (!matcher.find()) {
+        Matcher matcher = pattern.matcher(source.getBytes());
+        validateGroup(group, matcher.getEagerRegion());
+        int iGroup = Ints.checkedCast(group);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: same comment

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/RegexpFunctionsBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                    ThreadLocalRandom.current().ints(47, 58).limit(sourceLength).forEach(sliceOutput::appendByte);
+                    break;
+                case "literal":
+                    pattern = RegexpFunctions.castToRegexp(Slices.utf8Slice("literal"));
+                    // 97-122: 'a'-'z'
+                    ThreadLocalRandom.current().ints(97, 123).limit(sourceLength).forEach(sliceOutput::appendByte);
+                    break;
+            }
+            source = sliceOutput.slice();
+            Preconditions.checkState(source.length() == sourceLength, "source.length=%s, sourceLength=%s", source.length(), sourceLength);
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/RegexpFunctionsBenchmark.java]: static import

**Comment by cberner** [comment]
Looks good

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java (hunk 2 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: '}' byte can't be inside another codepoint

**Pull request merged**